### ISSUE TITLE
Negative Missing/Unexpected Values in Results (PySpark Pipeline)

### ISSUE THREAD

**Issue by alexsherstinsky** [fivetran/great_expectations#1133]
Hello:

The following problem occurred while running a PySpark pipeline on a CSV file.

Conditions:

1) PySpark Pipeline
2) Check DataFrame was not cached -- so all calculations had to be performed (due to the "lazy" mode when defining the pipeline operations).
3) Part of the Result ("SUMMARY" mode):
 ```
{
      "exception_info": {
        "exception_message": null, 
        "exception_traceback": null, 
        "raised_exception": false
      }, 
      "meta": {}, 
      "expectation_config": {
        "meta": {}, 
        "expectation_type": "expect_column_values_to_be_between", 
        "kwargs": {
          "column": "Week", 
          "max_value": 52, 
          "min_value": 1, 
          "mostly": null, 
          "result_format": "SUMMARY"
        }
      }, 
      "success": false, 
      "result": {
        "partial_unexpected_index_list": null, 
        "unexpected_count": -3, 
        "unexpected_percent": -0.003993291270665282, 
        "partial_unexpected_list": [], 
        "missing_percent": -0.005324388360887043, 
        "partial_unexpected_counts": [], 
        "element_count": 75126, 
        "unexpected_percent_nonmissing": -0.003993078663649674, 
        "missing_count": -4
      }
    }, 
```
The negative numbers are the issue (e.g., "missing_count" has to be a positive integer or 0).
4) This issue did not occur when the same pipeline was rerun with the only difference that the Check DataFrame in question was cached (using df_check.persist() in PySpark).

Thank you.

**Comment by jcampbell** [comment]
Thank you very much for documenting this, @alexsherstinsky.

After some digging, I believe we are experiencing an issue with a mismatch of caching behavior here: the dataframe was *not* cached in spark, but caching *was* enabled in Great Expectations. As a result, when the data changed in the underlying dataframe, the *correct* value of "missing_count" changed, but a stale value was returned by GE.

I think the immediate workarounds are twofold:
1. We should add a guard in GE that detects the obvious manifestations of this error (e.g. when missing or unexpected counts are *negative* and raises something like an `InvalidCacheValueError`

2. Ensure that we provide better messaging around when caching is or is not used. For example, in the `SqlAlchemyDataset`, similar issues are possible when operating on a table that could be changing.

**Comment by Aylr** [comment]
Fixed in #1152
